### PR TITLE
[Visit] Removed redundancy with project_rel table and standardize tables

### PR DESCRIPTION
**Pull request by ridz1208**
### Brief summary of changes
This PR mainly removes the `ProjectID` column in the new visit table. the project is redundant with the `project_rel` table and given the future direction of the Projects in LORIS, its easier to reinforce that relationship now rather then later.

other changes here:
 - update to reflect latest [standards](https://github.com/aces/Loris/blob/minor/docs/SQLModelingStandard.md)
 - updated RB, tests and controller

### CODE REVIEW

**Comment by ridz1208** [comment]
@PapillonMcGill Sorry, I know I instructed you to have the project in there but since then we had multiple discussions on reinforcing the projects role in loris.

ma bad

File: SQL/0000-00-00-schema.sql [hunk 1]
```diff
@@ -2140,20 +2140,17 @@ CREATE TABLE `candidate_consent_history` (
 CREATE TABLE `visit` (
   `VisitID` int(10) unsigned NOT NULL auto_increment,
   `VisitName` varchar(100) NOT NULL,
-  CONSTRAINT `visit_PK` PRIMARY KEY (`VisitID`),
-  CONSTRAINT `visit_name_UK` UNIQUE KEY (`VisitName`)
+  CONSTRAINT `PK_visit` PRIMARY KEY (`VisitID`),
+  CONSTRAINT `UK_visit_name` UNIQUE KEY (`VisitName`)
```
Comment (driusan) [inline SQL/0000-00-00-schema.sql]: what if the visit isn't in the UK?

Comment (ridz1208) [reply SQL/0000-00-00-schema.sql]: Im not sure if i should laugh or answer seriously

**Comment by PapillonMcGill** [comment]
approved after Travis agree!

**Comment by ridz1208** [comment]
@PapillonMcGill thx for reviewing. I will be making one more structural change that will actually be a half way between what I have now and what you had before.

Dismissing the review for now

**Comment by ridz1208** [comment]
@PapillonMcGill please review now

File: test/unittests/VisitTest.php [hunk 2]
```diff
@@ -76,112 +76,28 @@ protected function setUp()
...
-                                       2,
-                                       4,
-                                      ),
-                                      array(
-                                       $v6,
-                                       3,
-                                       3,
-                                      ),
-                                     );
+            array($v1, 1, 1),
```
Comment (johnsaigle) [inline test/unittests/VisitTest.php]: lol the old code here makes me want to die. Excellent change

**Comment by ridz1208** [comment]
@driusan I dont think anything is using this yet so I'm not sure if anything is testable other than the SQL patch

@sruthymathew123 @nicolasbrossard could you try running the SQL patch on IBIS and see if you get any errors?

File: SQL/New_patches/2018-01-17-normalisation_visit.sql [hunk 3]
```diff
@@ -1,25 +1,27 @@
...
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
 
 CREATE TABLE `visit_project_subproject_rel` (
+  `VisitProjectSubprojectID` int(10) unsigned NOT NULL AUTO_INCREMENT,
   `VisitID` int(10) unsigned NOT NULL,
-  `ProjectID` int(2) NOT NULL,
-  `SubprojectID` int(10) unsigned NOT NULL,
-  CONSTRAINT visit_project_subproject_rel_PK PRIMARY KEY (`VisitID`, `ProjectID`, `SubprojectID`),
-  CONSTRAINT visit_project_subproject_rel_VisitID_visit_VisitID_FK FOREIGN KEY (`VisitID`) 
+  `ProjectSubprojectID` int(10) unsigned NOT NULL,
```
Comment (driusan) [inline SQL/New_patches/2018-01-17-normalisation_visit.sql]: I find the name of this column confusing. Can we call it something like a "ProjectSubprojectPair"?

Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: I dont like that
`ProjectSubprojectTuple` ??

Comment (driusan) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: I'm fine with "Tuple" too.

Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: Made the change and added an amendment to the SQL modeling standards since that breaks the rules

https://github.com/aces/Loris/pull/4493/files#diff-190c4f51875cb53f0ebafdc6ae74b336R65

File: SQL/New_patches/2018-01-17-normalisation_visit.sql [hunk 4]
```diff
@@ -1,25 +1,27 @@
+ALTER TABLE project_rel DROP PRIMARY KEY, ADD COLUMN `ProjectSubprojectTupleID` int(10) unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT FIRST;
```
Comment (PapillonMcGill) [inline SQL/New_patches/2018-01-17-normalisation_visit.sql]: @ridz1208 Have you tested this line?
Had to split in 2 alter table lines in the past (1-drop PK, 2- add column with PK)

File: docs/SQLModelingStandard.md [hunk 5]
```diff
@@ -62,6 +62,8 @@ infrastructure will be re-designed to follow this guideline.*
   primary key should be `MyTableID`).
   - *i.e. the name of the primary identifier of the `country` table should be 
   `CountryID`.*
+> If a relational table requires a primary key ID (instead of the usual composite
```
Comment (PapillonMcGill) [inline docs/SQLModelingStandard.md]: I don't see the need for this.  <TableNameID> should be enough. 
This modification should be in a separate PR.
It should also be below the next point if you want to complement it. You now have 2 contradicting points (Line 65 and 67)

File: SQL/New_patches/2018-01-17-normalisation_visit.sql (hunk 4 above)
Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: I tested it on mysql 5.7

I actually could not split it because mysql refuses to drop a primary key unless you specify another one in the same statement

File: docs/SQLModelingStandard.md (hunk 5 above)
Comment (ridz1208) [reply docs/SQLModelingStandard.md]: Without this statement, the name of my primary key does not respect this document. it should be `ProjectSubprojectRelID` but I have `ProjectSubprojectTupleID`

And I dont see the contradiction with line 67, this line says that rel tables specifically should name their keys as such. line 67 says that most tables should have an autoincrement primary key

Comment (PapillonMcGill) [reply docs/SQLModelingStandard.md]: Sorry, contradict lines 59-64.

Does change my comment about should be in a separate PR. If we create exceptions in the standard every time someone doesn't like the standard, why have standards in the first place?

Comment (ridz1208) [reply docs/SQLModelingStandard.md]: I see this less as a contradiction and more as a specification, line 59 applies to all tables EXCEPT rel tables.

And not sure I would say this is trying to bypass the standards, I see this as amending the standards to include a usecase we had not thought of before... so more like complementing it and standards are not set in stone they should be updated if need be.

As far as the suggestion itself initially made by @driusan I kind of agree with the fact that `ProjectSubprojectTupleID` is clearer than `ProjectSubprojectID` or `ProjectSubprojectRelID` so the amendment here would be actually improving our standards.

What do you think ??

Comment (PapillonMcGill) [reply docs/SQLModelingStandard.md]: I don't see any 'EXCEPT' in the standard, so 'ProjectSubprojectRelID' is in line with the current standard.  If you want to modify/complement/improve the standard, it should be on a separate PR where people can study and comment appropriately.

Comment (ridz1208) [reply docs/SQLModelingStandard.md]: So in this PR I add a new column called `ProjectSubprojectRelID` wait for that to get merged and then open another one changing it to `tuple` ?

Comment (PapillonMcGill) [reply docs/SQLModelingStandard.md]: create a PR to modify the standard, once approved (if approved as nobody can predict the future, I surely can't) then use whatever the standard say.

Comment (ridz1208) [reply docs/SQLModelingStandard.md]: ORRRRR just call it rel

I pick option number 2

**Comment by ridz1208** [comment]
@PapillonMcGill I made the changes, please re-review carefully. due to the number of times changes were made it got messy

File: SQL/0000-00-00-schema.sql [hunk 6]
```diff
@@ -24,9 +24,11 @@ INSERT INTO subproject (title, useEDC, WindowDifference) VALUES
   ('Experimental', false, 'optimal');
 
 CREATE TABLE `project_rel` (
+  `ProjectSubprojectRelID` int(10) unsigned NOT NULL AUTO_INCREMENT,
```
Comment (PapillonMcGill) [inline SQL/0000-00-00-schema.sql]: @ridz1208 Shouldn't it be ProjectRelID?
unless you also want to rename the table to 'project_subproject_rel'.
don't forget to adjust PK def if needed and subsequent code.

File: SQL/New_patches/2018-01-17-normalisation_visit.sql [hunk 7]
```diff
@@ -1,25 +1,27 @@
+ALTER TABLE project_rel DROP PRIMARY KEY, ADD COLUMN `ProjectSubprojectRelID` int(10) unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT FIRST;
```
Comment (PapillonMcGill) [inline SQL/New_patches/2018-01-17-normalisation_visit.sql]: some version of MySQL will complain.
```suggestion
ALTER TABLE project_rel DROP PRIMARY KEY;
ALTER TABLE project_rek ADD COLUMN `ProjectSubprojectRelID` int(10) unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT FIRST;
```

File: SQL/0000-00-00-schema.sql (hunk 6 above)
Comment (ridz1208) [reply SQL/0000-00-00-schema.sql]: it's already done in another PR. I just didnt wanna do the changes twice or mix the code

File: SQL/New_patches/2018-01-17-normalisation_visit.sql (hunk 7 above)
Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: your suggestion gives me an error on mariadb.

Comment (driusan) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: What's the error? The deleted line (trying to drop and add in the same command) looks weird to me.

Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]:
```
Query execution failed

Reason:
SQL Error [1075] [42000]: (conn:228) Incorrect table definition; there can be only one auto column and it must be defined as a key
Query is : ALTER TABLE project_rel DROP PRIMARY KEY
```

File: SQL/0000-00-00-schema.sql (hunk 6 above)
Comment (PapillonMcGill) [reply SQL/0000-00-00-schema.sql]: Ok. reason accepted

File: SQL/New_patches/2018-01-17-normalisation_visit.sql (hunk 7 above)
Comment (PapillonMcGill) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: The error is not with the DROP statement, it is in the order of the parameter of the ADD COLUMN.  
`PRIMARY KEY` as to be after the `AUTO_INCREMENT`.

I haven't check well the ordering in your instruction.

Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: but if you see the error message. I get the error when I run the `ALTER TABLE project_rel DROP PRIMARY KEY` query

Comment (PapillonMcGill) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: Where you trying the entire drop and add in one statement?

Could you post the exact query and result here?

on my VM:
```
mysql> ALTER TABLE project_rel DROP PRIMARY KEY;
Query OK, 6 rows affected (0.46 sec)
Records: 6  Duplicates: 0  Warnings: 0
mysql> ALTER TABLE project_rel ADD COLUMN `ProjectSubprojectRelID` int(10) unsigned NOT NULL  AUTO_INCREMENT PRIMARY KEY FIRST;
Query OK, 0 rows affected (0.10 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]:
```
mysql> ALTER TABLE project_rel DROP PRIMARY KEY;
SQL Error [1075] [42000]: (conn:228) Incorrect table definition; there can be only one auto column and it must be defined as a key
```

Comment (PapillonMcGill) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: Can you post the current definition of you table?
```
SHOW CREATE TABLE project_rel
```

Comment (ridz1208) [reply SQL/New_patches/2018-01-17-normalisation_visit.sql]: NEVERMINDDDDD MA BAD
